toolchain/lexer/tokenized_buffer.cpp

```diff
@@ -692,6 +730,82 @@ class TokenizedBuffer::Lexer {
                        .column = current_column_});
   }
 
+  constexpr static auto MakeDispatchTable() -> DispatchTableT {
+    DispatchTableT table = {};
+    auto dispatch_lex_error = +[](Lexer& lexer, llvm::StringRef& source_text) {
+      return lexer.LexError(source_text);
+    };
+    for (int i = 0; i < 256; ++i) {
+      table[i] = dispatch_lex_error;
+    }
+
+    // First, set the first character of each symbol token spelling to dispatch
+    // to the symbol lexer. We use a `Period` placeholder for the token as there
+    // may be several different tokens that start with the same spelling. When
+    // that placeholder token kind is used, the symbol lexer will compute the
+    // exact symbol token kind.
+    auto dispatch_lex_symbol = +[](Lexer& lexer, llvm::StringRef& source_text) {
+      return lexer.LexSymbolToken(source_text, TokenKind::Period);
+    };
+#define CARBON_SYMBOL_TOKEN(TokenName, Spelling) \
+  table[(Spelling)[0]] = dispatch_lex_symbol;
+#include "toolchain/lexer/token_kind.def"
+
+    // Now special cased single-character symbols that are guaranteed to not
+    // join with another symbol. These are grouping symbols, terminators,
+    // or separators in the grammar and have a good reason to be
+    // orthogonal to any other punctuation. We do this separately because this
+    // needs to override soe of the generic handling above.
+#define CARBON_ONE_CHAR_SYMBOL_TOKEN(TokenName, Spelling)                  \
+  table[(Spelling)[0]] = +[](Lexer& lexer, llvm::StringRef& source_text) { \
+    return lexer.LexSymbolToken(source_text, TokenKind::TokenName);        \
+  };
+#include "toolchain/lexer/token_kind.def"
+
+    // For identifiers, keywords, and numeric type literals, we use the
+    // `Identifier` placeholder token kind.
+    auto dispatch_lex_word = +[](Lexer& lexer, llvm::StringRef& source_text) {
+      return lexer.LexKeywordOrIdentifier(source_text);
+    };
+    table['_'] = dispatch_lex_word;
+    // Note that we use raw loops because this needs to be `constexpr`
+    // evaluated.
```

Comment:
I assume this means "as opposed to `llvm::seq`" -- my first thought was that you were contrasting to
```c++
for (unsigned char c : "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ_") {
  table[c] = dispatch_lex_word;
}
```
or something like that.
```suggestion
    // Note that we don't use `llvm::seq` because this needs to be `constexpr`
    // evaluated.
```